### PR TITLE
add working agent_auth to windows

### PR DESCRIPTION
**Pull request by gaelmuller**
Add agent_auth functionnality to Windows.

Note that this is the same level of functionnality currently provided on UNIX: there is no authentication, resulting in poor security.

---
This change is [Reviewable](https://reviewable.io/reviews/ossec/ossec-hids/181)

### CODE REVIEW

**Comment by tkent** [comment]
This feature would have saved me several hours over the past few days. Would really be great to see it merged in.

**Comment by aquerubin** [comment]
The third parameter to OS_ConnectTCP() should be dropped.

**Comment by Ozonni** [comment]
What is the status of this PR?

**Comment by ddpbsd** [comment]
@Ozonni IIRC, the changes in this PR don't work with the current tree. I think I had some changes that would enable this feature for the win32 port, but I don't remember the status.

**Comment by MagnusMWW** [comment]
For anyone interested in building just the agent-auth.exe from the win-auth branch, you can just install a VM with CentOS6 or 7 and do the following (Other packages might be required, I did this on a CentOS6 with most development tools pre-installed):

>yum install epel-release

>yum install git mingw64-gcc

>git clone https://github.com/SekoiaLab/ossec-hids

>cd ossec-hids

>git checkout -b win-auth origin/win-auth

>cd src/win32

>./gen-win.sh

>cd ../win-pkg

>/usr/bin/x86_64-w64-mingw32-gcc -o agent-auth.exe -Wall  -DARGV0=\"agent-auth\" -DCLIENT -DWIN32 os_auth/agent_auth.c shared/*.c os_regex/*.c os_net/*.c os_crypto/md5/*.c win_service.c addagent/validate.c -Iheaders/ -I./ -lwsock32 -lsecur32

This should produce a working agent-auth.exe, without the dependency on specific directories or anything. Tested it, and you can run the tool stand-alone from anywhere, just copy the client.keys file to the ossec installation directory afterwards and it should be fine.

This might not be an option for everyone, but since we are re-packaging the OSSEC-agent in an MSI in my organization anyway, this turned out to be quite a nice solution for us. As @gaelmuller says, also remember that this version of the agent-auth.exe does not support any kind of authentication.

**Comment by soapergem** [comment]
Is this code ever going to be merged in? I found this from [a comment](http://serverfault.com/q/744116) on Server Fault. I want to automate the installation of OSSEC clients on Windows machines and it seems I need this agent-auth executable in order to do that. Alternatively, does anyone have an already compiled copy they could share?

**Comment by soapergem** [comment]
Also, I should mention that I tried following MagnusMWW's instructions (note that you have to add asterisks where Github added italics), and was able to compile the agent-auth.exe file, but when I run it, it just spits out this message: "This version of agent-auth.exe is not compatible with the version of Windows you're running. Check your computer's system information and the contact the software publisher."

**Comment by ddpbsd** [comment]
Unfortunately it requires time and energy. Both are in short supply,
neither are free.

**Comment by MagnusMWW** [comment]
@soapergem; Are possibly trying to run the 64-bit binary that the instructions i posted will build on a 32-bit Windows system? You can adjust the process to instead build a 32-bit binary by substituting:
`yum install git mingw64-gcc`
with
`yum install git mingw32-gcc`
and
`/usr/bin/x86-w32-mingw32-gcc -o agent-auth.exe -Wall -DARGV0=\"agent-auth\" -DCLIENT -DWIN32 os_auth/agent_auth.c shared/*.c os_regex/*.c os_net/*.c os_crypto/md5/*.c win_service.c addagent/validate.c -Iheaders/ -I./ -lwsock32 -lsecur32`
with
`/usr/bin/i686-w64-mingw32-gcc -o agent-auth.exe -Wall -DARGV0=\"agent-auth\" -DCLIENT -DWIN32 os_auth/agent_auth.c shared/*.c os_regex/*.c os_net/*.c os_crypto/md5/*.c win_service.c addagent/validate.c -Iheaders/ -I./ -lwsock32 -lsecur32`

btw, I'm updating my reply above to use code markdown instead of the simple ">", to avoid those starts being stripped out and confusing anyone else. Good catch!

**Comment by snaow** [comment]
Hi guys,

I have being testing everything and works fine on 2.8.3, but it does not on ossec-hids/master.
I made it work on Wazuh fork, I think it is working properly, I also create new rules in the Makefile so we can generate the Windows installer which include _agent-auth.exe._
Still pending implement the authentication.

Use development branch and run:
> cd src && make TARGET=winagent

The installer _ossec-win32-agent.exe_ generated is located on src/win32 folder.

Many thanks to @gaelmuller, @MagnusMWW  and everyone here.

**Comment by ddpbsd** [comment]
I got agent_auth.exe to compile, but it doesn't appear to do anything. Unfortunately I don't know how to do much debugging in windows, so any help would be appreciated.

**Comment by MagnusMWW** [comment]
@ddpbsd: That's strange... What system did you try to run it on? I only have a Windows 7 Pro (x86_64) systems accessible to test on, but both the 32-bit and the 64-bit build works for me on those systems. Afraid that I'm not that seasoned in Windows application debugging myself either, sorry.

If you'd like, you could try my builds and see if they work diffrently (they shouldn't really, but maybe worth a try?).

ftp://kalasarn.se/pub/temp/agent-auth-x86.exe
ftp://kalasarn.se/pub/temp/agent-auth-x86_64.exe

When running, i just have to use these options when invoking: `agent-auth-x86 -m <managerip>`

**Comment by soapergem** [comment]
@MagnusMWW I tried downloading the files you posted on your FTP site, but the file transfers consistently fail. If it's not too much trouble, can you either check on that, or share them a different way? Thanks so much.

**Comment by snaow** [comment]
I uploaded a working agent-auth.exe, tested on Win10, Win7 and Win 2012 R2.

[http://52.51.47.120:8000/agent-auth.exe](http://52.51.47.120:8000/agent-auth.exe)
([Virustotal URL check](https://www.virustotal.com/url/0a7164dc4b82e0cd18230ee183523e155ae358bfd80b23ef2785ea91a2bf7330/analysis/1473662974/))

**Comment by MagnusMWW** [comment]
@soapergem i see that @snaow beat me to it, but I've fixed the files now as well. Wrong permissions... ;)

**Comment by snaow** [comment]
Just trying to help @MagnusMWW :D 

btw we also included "-P" option for using ossec-authd password protected https://github.com/wazuh/ossec-wazuh/commit/6c7326d5dfd5bf3a60db5a5a98ea9e99205aa1c8

Still everything on development! WIP!

PS: I am using the agent-auth.exe daily and it works really good, you don't know how happy I am registering agents on Windows with no external-tools needed haha

**Comment by MagnusMWW** [comment]
@snaow: I appreciate it! :)

The password-protection sounds promising. Think I'll give that a go. Thanks for the tip!

It really helps, doesn't it? We have a tricky situation where we need to deploy a lot of agents on Windows machines during tight change-windows, and being able to use agent-auth certainly is helping with that! :)

**Comment by soapergem** [comment]
So I've noticed that this Windows version of the agent-auth utility doesn't support the -v flag to supply a CA certificate. But I also see some comments back and forth about native SSL vs. OpenSSL. I confess I'm a novice when it comes to certificate signing, so can someone tell me: does this (windows) agent-auth utility handle use of a certificate during authentication with the ossec-authd service? Even if it's just adding a certificate to Windows' store... would that work?

**Comment by igorgatis** [comment]
How long until this gets merged and make into a release?

File: src/win32/make.sh [hunk 1]
```diff
@@ -20,6 +20,7 @@ ${MING_BASE}-gcc -o ossec-agent.exe -Wall  -DARGV0=\"ossec-agent\" -DCLIENT -DWI
 ${MING_BASE}-gcc -o ossec-agent-eventchannel.exe -Wall  -DARGV0=\"ossec-agent\" -DCLIENT -DWIN32 -DOSSECHIDS -DEVENTCHANNEL_SUPPORT icon.o os_regex/*.c os_net/*.c os_xml/*.c zlib-1.2.8/*.c config/*.c shared/*.c os_execd/*.c os_crypto/blowfish/*.c os_crypto/md5/*.c os_crypto/sha1/*.c os_crypto/md5_sha1/*.c os_crypto/shared/*.c rootcheck/*.c *.c -Iheaders/ -I./ -lwsock32 -lwevtapi
 ${MING_BASE}-gcc -o ossec-rootcheck.exe -Wall  -DARGV0=\"ossec-rootcheck\" -DCLIENT -DWIN32 icon.o os_regex/*.c os_net/*.c os_xml/*.c config/*.c shared/*.c win_service.c rootcheck/*.c -Iheaders/ -I./ -lwsock32
 ${MING_BASE}-gcc -o manage_agents.exe -Wall  -DARGV0=\"manage-agents\" -DCLIENT -DWIN32 -DMA os_regex/*.c zlib-1.2.8/*.c os_zlib.c shared/*.c os_crypto/blowfish/*.c os_crypto/md5/*.c os_crypto/shared/*.c addagent/*.c -Iheaders/ -I./ -lwsock32 -lshlwapi
+${MING_BASE}-gcc -o agent-auth.exe -Wall  -DARGV0=\"agent-auth\" -DCLIENT -DWIN32 os_auth/agent_auth.c shared/*.c os_regex/*.c os_net/*.c os_crypto/md5/*.c win_service.c addagent/validate.c -Iheaders/ -I./ -lwsock32 -lsecur32
```
Comment (igorgatis) [inline src/win32/make.sh]: `os_auth/agent_auth.c` -> `agent_auth.c`
`win_service.c` -> `win32/win_service.c`

**Comment by stevenmdatt** [comment]
Is agent-auth now working correctly for automation purposes on windows?

Also @snaow your link doesn't work anymore, just FYI.

**Comment by snaow** [comment]
@stevenmdatt agent-auth is working properly in Wazuh fork, I've been using it every day since a few months ago.

You can download the release .msi and extract the agent-auth.exe tool from there:

- https://github.com/wazuh/wazuh/releases/download/v3.0.0/wazuh-agent-3.0.0-1.msi

It should be compatible with OSSEC 2.8/2.9.

**Comment by ddpbsd** [comment]
There's an `agent-auth.exe` created when compiling the windows version. Does it work?

**Comment by JLeClerc** [comment]
+1 to this PR. It would be very valuable to have this merged in.

**Comment by ddpbsd** [comment]
@JLeClerc Does it work?

**Comment by atomicturtle** [comment]
Closing this one out since we finished implementing this in another PR. Thanks for everyones contributions on this PR. Much appreciated everyone!

You can find this functionality in OSSEC 3.0.0, with enhancements already on the way for 3.0.1

**Pull request closed without merging**